### PR TITLE
apiserver: wrap errors in admission with context

### PR DESCRIPTION
**Pull request by stevekuznetsov**
When the API server encounters an error during admission webhook
handling, lower-level errors are bubbled up without any additional
context added. This leads to fairly opaque and unintelligible errors. It
is not clear to users if the API server itself is having an error (for
instance, fetching the REST client) or if the request to the webhook
failed in some way.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

/kind cleanup

```release-note
NONE
```

### CODE REVIEW

**Comment by sttts** [comment]
/assign @liggitt 

Sgtm.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go [hunk 1]
```diff
@@ -253,13 +253,13 @@ func (a *mutatingDispatcher) callAttrMutatingHook(ctx context.Context, h *admiss
...
 	if err := r.Do(ctx).Into(response); err != nil {
-		return false, &webhookutil.ErrCallingWebhook{WebhookName: h.Name, Reason: err}
+		return false, &webhookutil.ErrCallingWebhook{WebhookName: h.Name, Reason: fmt.Errorf("could not send webhook request: %w", err)}
 	}
 	trace.Step("Request completed")
 
 	result, err := webhookrequest.VerifyAdmissionResponse(uid, true, response)
 	if err != nil {
-		return false, &webhookutil.ErrCallingWebhook{WebhookName: h.Name, Reason: err}
+		return false, &webhookutil.ErrCallingWebhook{WebhookName: h.Name, Reason: fmt.Errorf("recieved invalid webhook response: %w", err)}
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:262]:
```suggestion
		return false, &webhookutil.ErrCallingWebhook{WebhookName: h.Name, Reason: fmt.Errorf("received invalid webhook response: %w", err)}
```

Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:262]: received

**Review by liggitt — COMMENTED** [review]
also update this to wrap with %w:

https://github.com/kubernetes/kubernetes/blob/21ab9c417562223bfe3329830788bc05ee29fb3f/staging/src/k8s.io/apiserver/pkg/util/webhook/error.go#L33-L38

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go [hunk 2]
```diff
@@ -253,13 +253,13 @@ func (a *mutatingDispatcher) callAttrMutatingHook(ctx context.Context, h *admiss
 	}
 
 	if err := r.Do(ctx).Into(response); err != nil {
-		return false, &webhookutil.ErrCallingWebhook{WebhookName: h.Name, Reason: err}
+		return false, &webhookutil.ErrCallingWebhook{WebhookName: h.Name, Reason: fmt.Errorf("could not send webhook request: %w", err)}
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:256]: `failed to call webhook at %s` with the URL

**Comment by liggitt** [comment]
did you sweep to see if there are places we are type-checking ErrCallingWebhook, unwrapping, and propagating the inner error in a way this wrapping would break?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go (hunk 2 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:256]: -1 on exposing webhook config details to the API caller

Comment (stevekuznetsov) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:256]: We expose the webhook URL in almost all other cases:

```
I0427 09:13:00.012155       1 build_controller.go:1779] Giving up retrying skuznets/cli-jq: failed to create build pod: Internal error occurred: failed calling webhook "pod-scaling.ci.openshift.io": Post "https://pod-scaler-consumer-admission.ci.svc:443/?timeout=5s": x509: certificate signed by unknown authority
I0427 15:13:00.100467       1 build_controller.go:1779] Giving up retrying skuznets/cli-jq: failed to create build pod: Internal error occurred: failed calling webhook "pod-scaling.ci.openshift.io": Post "https://pod-scaler-consumer-admission.ci.svc:443/?timeout=5s": no endpoints available for service "pod-scaler-consumer-admission"
I0427 15:28:38.317704       1 build_controller.go:1779] Giving up retrying skuznets/cli-jq: failed to create build pod: Internal error occurred: failed calling webhook "pod-scaling.ci.openshift.io": Post "https://pod-scaler-consumer-admission.ci.svc:443/?timeout=5s": context deadline exceeded
```

Does it matter much @liggitt  ?

**Comment by stevekuznetsov** [comment]
@liggitt  no, but with `%w` and `errors.Is` that should work fine, right? Or do you mean literal type unwrapping

**Comment by liggitt** [comment]
> @liggitt no, but with `%w` and `errors.Is` that should work fine, right?

If all the places we do this use `errors.Is`, sure. We're wildly inconsistent throughout the codebase.

> Or do you mean literal type unwrapping

Yeah, just need to check if we're type-casting to *ErrCallingWebhook, and propagating the internal error anywhere that is doing type-casting or passing to `apierrors.Is$Type` helper methods

**Comment by stevekuznetsov** [comment]
From my search I see some casts from `err` to `*ErrCallingWebhook` but no consumers of that cast ever do anything with the internal error @liggitt 

I think this is ready for review as it is, please let me know if any other changes are necessary.

**Pull request merged**